lib/lower/ASTBuilder.cpp

```diff
@@ -0,0 +1,836 @@
+//===-- lib/lower/ASTBuilder.cc -------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "flang/lower/ASTBuilder.h"
+#include "flang/parser/dump-parse-tree.h"
+#include "flang/parser/parse-tree-visitor.h"
+#include <algorithm>
+#include <cassert>
+#include <utility>
+
+namespace Fortran::lower {
+namespace {
+
+/// The instantiation of a parse tree visitor (Pre and Post) is extremely
+/// expensive in terms of compile and link time, so one goal here is to limit
+/// the bridge to one such instantiation.
+class ASTBuilder {
+public:
+  ASTBuilder() : pgm{new AST::Program}, parents{pgm.get()} {}
+
+  /// Get the result
+  std::unique_ptr<AST::Program> result() { return std::move(pgm); }
+
+  template <typename A>
+  constexpr bool Pre(const A &) {
+    return true;
+  }
+  template <typename A>
+  constexpr void Post(const A &) {}
+
+  // Module like
+
+  bool Pre(const parser::Module &node) { return enterModule(node); }
+  bool Pre(const parser::Submodule &node) { return enterModule(node); }
+
+  void Post(const parser::Module &) { exitModule(); }
+  void Post(const parser::Submodule &) { exitModule(); }
+
+  // Function like
+
+  bool Pre(const parser::MainProgram &node) { return enterFunc(node); }
+  bool Pre(const parser::FunctionSubprogram &node) { return enterFunc(node); }
+  bool Pre(const parser::SubroutineSubprogram &node) { return enterFunc(node); }
+  bool Pre(const parser::SeparateModuleSubprogram &node) {
+    return enterFunc(node);
+  }
```

Comment:
Here, it seems the same function is called. Isn't it possible to write a single Pre function for all these?
Same question for the many following Post functions below.

Reply:
It is possible do to so, but as I answered to Peter, we do not want to condense the implementation too much based on today's pattern because we think the explicit Pre/Post makes it easy to reason about and work with for now.

Reply:
Eventually condensed as suggested.